rpc/server_interceptors.go

```diff
@@ -0,0 +1,126 @@
+package rpc
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/edaniels/golog"
+	"go.opencensus.io/trace"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+// UnaryServerTracingInterceptor starts a new Span if Span metadata exists in the context.
+func UnaryServerTracingInterceptor(logger golog.Logger) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		remoteSpanContext, err := remoteSpanContextFromContext(ctx)
+		var span *trace.Span
+		if err == nil {
+			ctx, span = trace.StartSpanWithRemoteParent(ctx, "server_root", remoteSpanContext)
+			defer span.End()
+		} else {
+			logger.Warnf("client did not send valid Span metadata in headers, local Spans will not be linked to client. reason: %w", err)
+		}
+
+		resp, err := handler(ctx, req)
+		if err == nil {
+			return resp, nil
+		}
+		if _, ok := status.FromError(err); ok {
+			return nil, err
+		}
+		if s := status.FromContextError(err); s != nil {
+			return nil, s.Err()
+		}
+		return nil, err
+	}
+}
+
+// StreamServerTracingInterceptor starts a new Span if Span metadata exists in the context.
+func StreamServerTracingInterceptor(logger golog.Logger) grpc.StreamServerInterceptor {
+	return func(srv interface{}, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		remoteSpanContext, err := remoteSpanContextFromContext(stream.Context())
+		if err == nil {
+			newCtx, span := trace.StartSpanWithRemoteParent(stream.Context(), "server_root", remoteSpanContext)
+			defer span.End()
+			stream = wrapServerStream(newCtx, stream)
+		} else {
+			logger.Warnf("client did not send valid Span metadata in headers, local Spans will not be linked to client. reason: %w", err)
+		}
+
+		err = handler(srv, stream)
+		if err == nil {
+			return nil
+		}
+		if _, ok := status.FromError(err); ok {
+			return err
+		}
+		if s := status.FromContextError(err); s != nil {
+			return s.Err()
+		}
+		return err
+	}
+}
+
+type serverStreamWrapper struct {
+	grpc.ServerStream
+	ctx context.Context
+}
+
+func (s *serverStreamWrapper) Context() context.Context {
```

Comment:
I think our linter requires all public methods to have documentation. You can check locally by running `make lint-go`. Nevertheless, it's good practice. I'd just copy the docs from the method you're overriding.